packages/dds/tree/src/tree/delta.ts

```diff
@@ -366,3 +371,110 @@ export function inputLength(mark: Mark): number {
         default: unreachableCase(type);
     }
 }
+
+/**
+ * Converts inserted content into the format expected in Delta instances.
+ * This involves applying the following changes:
+ * - Updating node values
+ * - Inserting new subtrees within the inserted content
+ * - Deleting parts of the inserted content
+ *
+ * The only kind of change that is not applied by this function is MoveIn.
+ *
+ * @param node - The subtree to apply modifications to. Updated in place.
+ * @param modify - The modifications to either apply or collect.
+ * @returns The remaining modifications that the consumer of the Delta will apply on the given node. May be empty if
+ *   all modifications are applied by the function.
+ */
+ export function applyModifyToInsert(
+    node: ProtoNode,
+    modify: Modify,
+): Map<FieldKey, MarkList<InsertedFieldsMark>> {
+    const outFieldsMarks: Map<FieldKey, MarkList<InsertedFieldsMark>> = new Map();
+    if (modify.setValue !== undefined) {
+        node.value = modify.setValue.value;
+    }
+    if (modify.fields !== undefined) {
+        const protoFields = node.fields ?? {};
+        const modifyFields = modify.fields;
+        for (const brandedKey of modifyFields.keys()) {
+            const key = brandedKey as string;
+            const outNodes = protoFields[key] ?? fail("Missing field");
+            const outMarks = new OffsetListFactory<InsertedFieldsMark>();
+            let index = 0;
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            for (const mark of modifyFields.get(brandedKey)!) {
+                if (isSkipMark(mark)) {
+                    index += mark;
+                    outMarks.pushOffset(mark);
+                } else {
+                    // Inline into `switch(mark.type)` once we upgrade to TS 4.7
+                    const type = mark.type;
+                    switch (type) {
+                        case MarkType.Insert: {
+                            outNodes.splice(index, 0, ...mark.content);
+                            index += mark.content.length;
+                            outMarks.pushOffset(mark.content.length);
+                            break;
+                        }
+                        case MarkType.InsertAndModify: {
```

Comment:
Don't we need to recurse into `applyModifyToInsert` in this case?

Reply:
I think we can assume that everything in the `Modify` is already normalized since that's the contract for deltas.

Reply:
Ah, that's right. That's an interesting shift from the changeset-based implementation in `toDelta.ts`.